samtranslator/model/sam_resources.py

```diff
@@ -1785,88 +1787,106 @@ def to_cloudformation(self, **kwargs: Any) -> List[Resource]:  # type: ignore
+                    )
 
-        verify_profile_variables_replaced(profile_properties)
+            replacement = {
+                "Source.Arn": source.arn,
+                "Destination.Arn": destination.arn,
+                "Source.ResourceId": source.resource_id,
+                "Destination.ResourceId": destination.resource_id,
+                "Source.Name": source.name,
+                "Destination.Name": destination.name,
+                "Source.Qualifier": source.qualifier,
+                "Destination.Qualifier": destination.qualifier,
+            }
 
-        generated_resources: List[Resource] = []
-        if profile_type == "AWS_IAM_ROLE_MANAGED_POLICY":
-            generated_resources.append(
-                self._construct_iam_policy(source, destination, profile_properties, resource_resolver)
-            )
-        if profile_type == "AWS_SQS_QUEUE_POLICY":
-            generated_resources.append(self._construct_sqs_queue_policy(source, destination, profile_properties))
-        if profile_type == "AWS_SNS_TOPIC_POLICY":
-            generated_resources.append(self._construct_sns_topic_policy(source, destination, profile_properties))
-        if profile_type == "AWS_LAMBDA_PERMISSION":
-            generated_resources.extend(
-                self._construct_lambda_permission_policy(source, destination, profile_properties)
-            )
+            try:
+                profile_properties = profile_replace(profile_properties, replacement)
+            except ValueError as e:
+                raise InvalidResourceException(self.logical_id, str(e))
+
+            verify_profile_variables_replaced(profile_properties)
+
+            generated_resources: List[Resource] = []
 
-        generated_logical_ids = [resource.logical_id for resource in generated_resources]
+            if profile_type == "AWS_IAM_ROLE_MANAGED_POLICY":
+                generated_resources.append(
+                    self._construct_iam_policy(
+                        source, destination, profile_properties, resource_resolver, index, multi_dest
+                    )
+                )
+            if profile_type == "AWS_SQS_QUEUE_POLICY":
+                generated_resources.append(
+                    self._construct_sqs_queue_policy(source, destination, profile_properties, index, multi_dest)
+                )
+            if profile_type == "AWS_SNS_TOPIC_POLICY":
+                generated_resources.append(
+                    self._construct_sns_topic_policy(source, destination, profile_properties, index, multi_dest)
+                )
+            if profile_type == "AWS_LAMBDA_PERMISSION":
+                generated_resources.extend(
+                    self._construct_lambda_permission_policy(source, destination, profile_properties, index, multi_dest)
+                )
+
```

Comment:
here we should do a `raise TypeError(f"Unknown profile policy type '{profile_type}'")`

Reply:
agree